workflow/rules/diffexp.smk

```diff
@@ -262,6 +242,26 @@ rule logcount_matrix:
         "../scripts/sleuth-to-matrix.R"
 
 
+rule plot_diffexp_heatmap:
+    input:
+        logcountmatrix_file="results/tables/logcount-matrix/{model}.logcount-matrix.tsv",
+    output:
+        diffexp_heatmap=report(
+            "results/plots/diffexp-heatmap/{model}.diffexp-heatmap.pdf",
+            caption="../report/plot-heatmap.rst",
+            category="Heatmaps",
+        ),
```

Comment:
This rule does not yet behave as I would have expected.
I would do it in the following way:
```suggestion
    output:
        diffexp_heatmap=report(
            "results/plots/diffexp-heatmap/{model}.diffexp-heatmap.{mode}.pdf",
            caption="../report/plot-heatmap.rst",
            category="Heatmaps",
        ),
    wildcard_constraints:
        mode="topn|predefined"
```
We have a wildcard for the mode (top n differentially expressed genes, or predefined list). Depending on the mode, the script should do the following:

* `snakemake@wildcards[["mode"]] == "topn"`: fill `selectedgenes` such that it selects the top 50 genes from the differential expression dataframe (sorted by p-value).
* `snakemake@wildcards[["mode"]] == "predefined": fill `selectedgenes` such that it selects the genes from the predefined list.

This is an if-else at the beginning. The rest of the script should be outside of that if-else, and just generate the heatmap based on the selectedgenes variable.

Reply:
Here, Since this is the final output, We need to define it in common.smk. In that case, I am a bit confused regarding the wildcard_constraints. The mode has to be dependent upon the input file `predef_genelist`. So whether the mode has to be defined in the function which checks if the file exists should return mode as `predefined` or else ` topn`?

Reply:
If the predefined genelist is given in the config, I want both heatmaps, otherwise, I only want the topn.

Reply:
So this is what you request then in the final output in common smk. And in the first case, the rule is executed twice (once with mode predefined, once with topn), in the other case, you request only topn in the final output, such that the rule is only run once.

Reply:
The rule now runs two times based on the predefined genelist set to `true` and once if it is set to `false` given in the config.yaml file.